jolt-core/src/r1cs/spartan.rs

```diff
@@ -473,57 +427,417 @@ where
+        self.input_claim
+    }
+
+    fn compute_prover_message(&self, _round: usize) -> Vec<F> {
+        let prover_state = self
+            .prover_state
+            .as_ref()
+            .expect("Prover state not initialized");
+        let degree = <Self as BatchableSumcheckInstance<F, ProofTranscript>>::degree(self);
+
+        let univariate_poly_evals: Vec<F> = (0..prover_state.unexpanded_pc_poly.len() / 2)
+            .into_par_iter()
+            .map(|i| {
+                let unexpanded_pc_evals = prover_state.unexpanded_pc_poly.sumcheck_evals(
+                    i,
+                    degree,
+                    BindingOrder::HighToLow,
+                );
+                let pc_evals =
+                    prover_state
+                        .pc_poly
+                        .sumcheck_evals(i, degree, BindingOrder::HighToLow);
+                let eq_evals = prover_state.eq_plus_one_poly.sumcheck_evals(
+                    i,
+                    degree,
+                    BindingOrder::HighToLow,
+                );
+
+                vec![
+                    (unexpanded_pc_evals[0] + prover_state.r * pc_evals[0]) * eq_evals[0], // eval at 0
+                    (unexpanded_pc_evals[1] + prover_state.r * pc_evals[1]) * eq_evals[1], // eval at 2
+                ]
+            })
+            .reduce(
+                || vec![F::zero(); degree],
+                |mut running, new| {
+                    for i in 0..degree {
+                        running[i] += new[i];
+                    }
+                    running
+                },
+            );
+
+        univariate_poly_evals
+    }
+
+    fn bind(&mut self, r_j: F, _round: usize) {
+        let prover_state = self
+            .prover_state
+            .as_mut()
+            .expect("Prover state not initialized");
+
+        rayon::join(
+            || {
+                rayon::join(
+                    || {
+                        prover_state
+                            .unexpanded_pc_poly
+                            .bind_parallel(r_j, BindingOrder::HighToLow)
+                    },
+                    || {
+                        prover_state
+                            .pc_poly
+                            .bind_parallel(r_j, BindingOrder::HighToLow)
+                    },
+                )
+            },
+            || {
+                prover_state
+                    .eq_plus_one_poly
+                    .bind_parallel(r_j, BindingOrder::HighToLow)
+            },
+        );
+    }
+
+    fn cache_openings(&mut self) {
+        let prover_state = self
+            .prover_state
+            .as_ref()
+            .expect("Prover state not initialized");
+
+        let unexpanded_pc_eval = prover_state.unexpanded_pc_poly.final_sumcheck_claim();
+        let pc_eval = prover_state.pc_poly.final_sumcheck_claim();
+        let eq_plus_one_eval = prover_state.eq_plus_one_poly.final_sumcheck_claim();
```

Comment:
don't need to cache this claim because the verifier evaluates `EqPlusOnePolynomial` itself